sstables_loader.cc

```diff
@@ -203,11 +205,142 @@ class tablet_sstable_streamer : public sstable_streamer {
         return result;
     }
 
-    future<> stream_fully_contained_sstables(const dht::partition_range& pr, std::vector<sstables::shared_sstable> sstables, shared_ptr<stream_progress> progress) {
-        // FIXME: fully contained sstables can be optimized.
+    struct minimal_sst_info {
+        sstables::generation_type _generation;
+        sstables::sstable_version_types _version;
+        sstables::sstable_format_types _format;
+    };
+    using sst_classification_info = std::vector<std::vector<minimal_sst_info>>;
+
+    future<>
+    stream_fully_contained_sstables(const dht::partition_range& pr, std::vector<sstables::shared_sstable> sstables, shared_ptr<stream_progress> progress) {
+        if (_stream_scope == stream_scope::node && _from_object_storage) {
+            llog.debug("Directly downloading {} fully contained SSTables to local node from object storage.", sstables.size());
+            return download_fully_contained_sstables(std::move(sstables), std::move(progress)).then([this](auto downloaded_ssts) -> future<> {
+                auto dwnld_ssts = std::move(downloaded_ssts);
+                auto dwnld_ssts_its = dwnld_ssts | std::views::transform([](auto& ssts) { return ssts.cbegin(); }) | std::ranges::to<std::vector>();
+                while (true) {
```

Comment:
Better to reformulate as a terminating loop rather than asking the reader to hunt for the termination condition.